Add inspector multi-select editing

dugramen opened the pull request:
Allows editing multiple properties in the inspector at the same time. Multiselecting is done with Shift+click, single select with Ctrl+click.

If you select multiple resources, you can edit the same property(s) on multiple resources at a time. Video example:


https://github.com/user-attachments/assets/0711e0f0-ebde-4b52-a016-55937ea3ae22



This should alleviate Stylebox and theming tedium described in proposals like https://github.com/godotengine/godot-proposals/issues/13034, https://github.com/godotengine/godot-proposals/issues/8094, https://github.com/godotengine/godot-proposals/issues/9743.

Commit:
Add inspector multi-select editing

chocola-mint commented:
Minor nitpick: In your demo, around the 0:11 mark, there's a frame where the other four properties aren't updated to show the correct color immediately. Your mouse then moved, triggering redraw, but it'd be nice if they were updated in real time.

Perhaps it has something to do with how ColorPickers are implemented?

dugramen commented:
@chocola-mint Yeah that's a color picker thing.The preview updates as you drag the color around, but the actual property doesn't set until you let go or close. I think it's to avoid any setter code running so many times. But you'll see the spin boxes updates as you drag, but still at like a low "frame rate"

Ivorforce commented:
Which value will be displayed in the inspector if there are different values across selected items? I think this is the crux of what makes this feature difficult to get right.

Most mature 'multi-selection' software I've seen will show a `<multiple values>` placeholder, or at worst, a unique "not set" state. 
This is important because it will prevent accidental changes to values that were assumed to be the same. Also, it can make 'unification' of values easier (if the value is already set for some items in the selection), because otherwise, there have to be two clicks to trigger a change across the whole selection.

dugramen commented:
Rebased and made it so this behavior is disabled for array items. Turns out when editing array items, the inspector doesn't just change that 1 item but sets the whole array again. So if multiselecting, entire arrays would be copied into each other rather than just the 1 item you might expect

Also updated the main post's video to show bulk multiselect with shift+click

AThousandShips commented on editor/inspector/editor_inspector.cpp:
```diff
@@ -1000,9 +1000,58 @@ void EditorProperty::select(int p_focusable) {
+			int i = 0;
+			bool selecting = false;
+			while (i < nodes.size()) {
+				Control *control = cast_to<Control>(nodes[i]);
```
```suggestion
				Control *control = Object::cast_to<Control>(nodes[i]);
```

AThousandShips commented on editor/inspector/editor_inspector.cpp:
```diff
@@ -1000,9 +1000,58 @@ void EditorProperty::select(int p_focusable) {
+			while (i < nodes.size()) {
+				Control *control = cast_to<Control>(nodes[i]);
+				if (control) {
+					EditorProperty *ep = cast_to<EditorProperty>(control);
```
```suggestion
					EditorProperty *ep = Object::cast_to<EditorProperty>(control);
```

AThousandShips commented on editor/inspector/editor_inspector.cpp:
```diff
@@ -1108,12 +1155,13 @@ void EditorProperty::gui_input(const Ref<InputEvent> &p_event) {
 		}
-		if (delete_rect.has_point(mpos)) {
+
+		else if (delete_rect.has_point(mpos)) {
```
Same here

AThousandShips commented on editor/inspector/editor_inspector.cpp:
```diff
@@ -5137,6 +5193,78 @@ void EditorInspector::_edit_request_change(Object *p_object, const String &p_pro
 	}
 }
 
+void EditorInspector::_process_prop_undo_redo(EditorUndoRedoManager *undo_redo, Object *obj, String prop_name, Variant p_value, bool p_refresh_all) {
```
```suggestion
void EditorInspector::_process_prop_undo_redo(EditorUndoRedoManager *p_undo_redo, Object *p_obj, const String &p_prop_name, const Variant &p_value, bool p_refresh_all) {
```

AThousandShips commented on editor/inspector/editor_inspector.cpp:
```diff
@@ -5137,6 +5193,78 @@ void EditorInspector::_edit_request_change(Object *p_object, const String &p_pro
+	Variant v_object = obj;
+	Variant v_name = prop_name;
+	const Vector<Callable> &callbacks = EditorNode::get_editor_data().get_undo_redo_inspector_hook_callback();
+	for (int i = 0; i < callbacks.size(); i++) {
```
For each
